### PR TITLE
Fix serialization of Params with set data type

### PR DESCRIPTION
**Pull request by dstandish**
This is a solution for https://github.com/apache/airflow/issues/19096

Previously, the serialization of params did not run the param value through the `_serialize` function, resulting in non-json-serializable dictionaries.  This manifested when a user, for example, tried to use params with a default value of type `set`.

Here we change the logic to run the param value through the serialization process.  And I add a test for the `set` case.

Could def use extra set of eyes on schema definition. btw i added hook for formatting this -- i can do that in separate pr if that's helpful.

### CODE REVIEW

File: airflow/serialization/serialized_objects.py [hunk 1]
```diff
@@ -409,16 +407,26 @@ def _value_is_hardcoded_default(cls, attrname: str, value: Any, instance: Any) -
...
+    def _serialize_param(cls, param: Param):
+        d = param.dump()
+        d['value'] = cls._serialize(d['value'])
+        return d
+
+    @classmethod
+    def _deserialize_param(cls, param: Dict):
+        param_class = import_string(param['__class'])
+        param['value'] = cls._deserialize(param['value'])
+        return param_class(default=param['value'], description=param['description'])
```
Comment (uranusjr) [inline airflow/serialization/serialized_objects.py]: How does this work with DAGs already serialised in the database? Do we need to implement some detection to ensure those entries can be deserialised correctly?

Comment (dstandish) [reply airflow/serialization/serialized_objects.py]: it's  a good point, have to look into that

Comment (dstandish) [reply airflow/serialization/serialized_objects.py]: ok in latest commit i think i fixed my fix, so that it should be no change to serialization structure.    initially i had used `value` but now i restore `default` as the key.  theres just ambiguity because `default` is the kwarg name, and `value` is the attr.

File: airflow/serialization/serialized_objects.py [hunk 2]
```diff
@@ -409,16 +407,26 @@ def _value_is_hardcoded_default(cls, attrname: str, value: Any, instance: Any) -
...
+    @classmethod
+    def _serialize_param(cls, param: Param):
+        param_dict = param.dump()
+        param_dict['default'] = cls._serialize(param_dict.pop('value'))
+        return param_dict
+
+    @classmethod
+    def _deserialize_param(cls, param_dict: Dict):
+        param_class = import_string(param_dict['__class'])
+        param_dict['default'] = cls._deserialize(param_dict['default'])
```
Comment (jedcunningham) [inline airflow/serialization/serialized_objects.py]: Should we be doing `_deserialize` on everything in `param_dict`?

Comment (dstandish) [reply airflow/serialization/serialized_objects.py]: yeah probably so...

Comment (dstandish) [reply airflow/serialization/serialized_objects.py]: ok updated

File: airflow/models/param.py [hunk 3]
```diff
@@ -49,6 +49,7 @@ class Param:
     """
 
     __NO_VALUE_SENTINEL = NoValueSentinel()
+    CLASS_IDENTIFIER = '__class'
```
Comment (uranusjr) [inline airflow/models/param.py:51]: I think we use `__type` in most places so this should also do for consistency.

Comment (dstandish) [reply airflow/models/param.py:51]: I think that would be a schema change.

Note that type is used for deciding how to serialize, but in this case, class is one layer below.  It's probably redundant and could possibly be combined but this is the existing structure

Comment (uranusjr) [reply airflow/models/param.py:51]: Oh! So the inconsistency is already there (in 2.2). In that case I guess it's better to stick to it.

**Review by uranusjr — DISMISSED** [review]

File: airflow/serialization/serialized_objects.py [hunk 4]
```diff
@@ -417,7 +417,10 @@ def _serialize_param(cls, param: Param):
     @classmethod
     def _deserialize_param(cls, param_dict: Dict):
         param_class = import_string(param_dict.pop(Param.CLASS_IDENTIFIER))
-        param_kwargs = cls._deserialize(param_dict)
+        try:
+            param_kwargs = cls._deserialize(param_dict)
+        except KeyError:
+            param_kwargs = param_dict
```
Comment (uranusjr) [inline airflow/serialization/serialized_objects.py]: What is causing the KeyError? It feels wrong to just "give up" on deserializing.

Comment (dstandish) [reply airflow/serialization/serialized_objects.py]: that's not what's going on here
the KeyError is essentially 2.2.0 serialiazation approach, which for params, doesn't run their attrs through our serialization logic, but assumes they are json serializable (which is the problem, because in the case of sets they are not)

but so when you try to _deserialize something that was not _serialize(d), you get a KeyError and in this case we can just take the args as they are -- they weren't "serialized" anyway, so we don't need to "deserialize" them

_however_, i am working on a better approach, cleaner, and i'll push it tomorrow.

Comment (dstandish) [reply airflow/serialization/serialized_objects.py]: separately though, do we really need to support non-json-serializable params?  this will cause problems, for example when you run trigger dag in the UI -- probably in that case they _must_ be valid json.  so what's the point of supporting non-json in the default?  i think it would be best if we simply require that params be valid json.  if users want to use set operations in their templates, they can do so in the manner i demonstrated in the reported issue; and we can still keep the param values themselves json. @jedcunningham @kaxil interested in your thougts too.

Comment (dstandish) [reply airflow/serialization/serialized_objects.py]: and another reason to be skeptical of this "fix" -- the big feature here is that params can have validation logic built-in.  but the validation approach is [json schema validation](https://json-schema.org/draft/2020-12/json-schema-validation.html), i.e. it explicitly assumes you are validating json!  so why would we allow params to not be json (or json-serializable).

Comment (dstandish) [reply airflow/serialization/serialized_objects.py]: after chatting with kaxil, we need to restore support for this before properly deprecating

File: .pre-commit-config.yaml [hunk 5]
```diff
@@ -195,6 +195,15 @@ repos:
           - "4"
         files: ^chart/values\.schema\.json$|^chart/values_schema\.schema\.json$
         pass_filenames: true
+      - id: pretty-format-json
+        name: Serialization schema
+        args:
+          - --autofix
+          - --no-sort-keys
+          - --indent
+          - "2"
```
Comment (kaxil) [inline .pre-commit-config.yaml]: Why is this separate than the above pre-commit and why `2` indent instead of `4`

Comment (kaxil) [reply .pre-commit-config.yaml]: I remember why we hadn't included this file yet in above pre-commit, we formatted in a way that was a bit easy to read.

Example:

image

vs

image


However no strong preference

Comment (kaxil) [reply .pre-commit-config.yaml]: In any case, let's separate the formatting and pre-commit change to a different PR than the bugfix please

File: airflow/serialization/serialized_objects.py [hunk 6]
```diff
@@ -325,7 +324,7 @@ def _serialize(cls, var: Any) -> Any:  # Unfortunately there is no support for r
         elif isinstance(var, TaskGroup):
             return SerializedTaskGroup.serialize_task_group(var)
         elif isinstance(var, Param):
-            return cls._encode(var.dump(), type_=DAT.PARAM)
+            return cls._encode(cls._serialize_param(var), type_=DAT.PARAM)
```
Comment (kaxil) [inline airflow/serialization/serialized_objects.py:327]: Thinking about this more: @msumit  do you think we will be able to validate `set` or non-JSON structures using the JSON Schema validation?

File: airflow/models/param.py (hunk 3 above)
Comment (msumit) [reply airflow/models/param.py:51]: In the original PR, I initially used `__type` only, but Ash felt that it should be called `__class`, cause it would be representing a Param class & not a native type, which made sense to me.

File: airflow/serialization/serialized_objects.py [hunk 7]
```diff
@@ -409,30 +406,59 @@ def _value_is_hardcoded_default(cls, attrname: str, value: Any, instance: Any) -
             return True
         return False
 
+    @classmethod
+    def _serialize_param(cls, param: Param):
+        return dict(
+            __class=f"{param.__module__}.{param.__class__.__name__}",
+            default=cls._serialize(param.value),
+            description=cls._serialize(param.description),
```
Comment (msumit) [inline airflow/serialization/serialized_objects.py:414]: Do we really need to serialize/deserialize a parma's description as well? It ought to be a simple string only.

**Comment by msumit** [comment]
lgtm overall. Also nice work in handling the backward compatibility 👍

**Review by msumit — APPROVED** [review]

File: airflow/serialization/serialized_objects.py (hunk 7 above)
Comment (dstandish) [reply airflow/serialization/serialized_objects.py:414]: So I think it ends up being the same because string is primitive type

Comment (dstandish) [reply airflow/serialization/serialized_objects.py:414]: yeah just double checked and when it is primitive type, `_serialize` doesn't do anything

though, maybe we should not serialize it so as not to accidentally support something users should not be doing anyway.  so on second thought, probably should change this to put the raw value  (in which case, json-serialization will fail)

Comment (dstandish) [reply airflow/serialization/serialized_objects.py:414]: ok just sat down to not `_serialize` the `description` attr.  and on second thought, i think it's better to run it through `_serialize` after all.  like, that's precisely what `_serialize` is there for.  there's a reason that `_serialize` has handling for primitive types, and so we should use it; to do otherwise is essentially to duplicate serialization logic.  also, not running through `_serialize` would not achieve the goal of preventing a user from using non-string values (it only would prevent from using non-json-serializable types such as `set`) -- if we really want to clamp down on types, probably the right place do do that is in `Param.__init__`.

so, i think i'll leave this as is, but let me know if you think otherwise.

thanks

File: .pre-commit-config.yaml (hunk 5 above)
Comment (dstandish) [reply .pre-commit-config.yaml]: yeah i'll split it out.

but yeah it was previously not processed in pre-commit.  but it should be to ensure consistent formatting, even if that means slightly less pretty.
and yeah i just chose indent of 2 because that's what the file already had.

File: airflow/serialization/serialized_objects.py (hunk 6 above)
Comment (dstandish) [reply airflow/serialization/serialized_objects.py:327]: can you clarify what you mean @kaxil ?   the serializer converts them to json...

File: .pre-commit-config.yaml (hunk 5 above)
Comment (dstandish) [reply .pre-commit-config.yaml]: ok updated

File: airflow/serialization/serialized_objects.py (hunk 6 above)
Comment (kaxil) [reply airflow/serialization/serialized_objects.py:327]: The following code runs jsonschema validation if one is supplied.

https://github.com/apache/airflow/blob/02b7e2c092902bf42791a5ef1a70bc71226f1e32/airflow/models/param.py#L47-L51

Example test:

https://github.com/apache/airflow/blob/02b7e2c092902bf42791a5ef1a70bc71226f1e32/tests/models/test_param.py#L58-L66

This won't be possible with "set" is what I mean:

i.e. the following works:

```python
    def test_set_param(self):
        p = Param({'a', 'b'})
        assert p.resolve() == {'a', 'b'}
```

but there is no way of adding validation like the set only contains <10 elements, example:

```python
    def test_set_param(self):
        p = Param({'a', 'b'}, type='set', maxItems=1)
        assert p.resolve() == {'a', 'b'}
```

**It is not a big deal but worth pointing that out in the docs** but that's a separate issue and can be in a separate PR

**Comment by kaxil** [comment]
@dstandish Can you take a look at the failing tests please

**Comment by dstandish** [comment]
> @dstandish Can you take a look at the failing tests please

yup will do

File: airflow/serialization/serialized_objects.py (hunk 6 above)
Comment (dstandish) [reply airflow/serialization/serialized_objects.py:327]: oh, _that_ kind of validation.  so yeah this is why i think we should immediately deprecate this feature (namely, support for non-json-serializable params), and perhaps warn whenever we get params with a default that is non-json-serializable.

do you think i should add such a warning to this PR?

File: tests/serialization/test_dag_serialization.py [hunk 8]
```diff
@@ -1433,29 +1461,32 @@ def test_serialized_objects_are_sorted(self, object_to_serialized, expected_outp
...
             "dag": {
                 "_dag_id": "simple_dag",
-                "fileloc": __file__,
+                "fileloc": '__file__',
                 "tasks": [],
                 "timezone": "UTC",
                 "params": {"none": None, "str": "str", "dict": {"a": "b"}},
             },
         }
-        SerializedDAG.validate_schema(serialized)
```
Comment (dstandish) [inline tests/serialization/test_dag_serialization.py:1446]: fyi @jedcunningham @kaxil @msumit we have to remove this line because it is validating pre-2.2.0 params against the current schema.  

it's not needed because `validate_schema` is only called when serializing; not when deserializing, and the test still verifies that the json deserializes properly.

if we wanted to keep this line, we'd have to loosen `schema.json` to allow arbitrary dict:

```json
"params": {
    "anyOf": [
        { "$ref": "#/definitions/params_dict" },
        {"$ref": "#/definitions/dict"}]
},
```

But this is no longer the serialization schema.

**Comment by dstandish** [comment]
fyi i updated Param init to check whether the `default` is json serializable and a deprecation warning if it is not.

File: tests/serialization/test_dag_serialization.py [hunk 9]
```diff
@@ -733,7 +735,10 @@ def test_dag_params_roundtrip(self, val, expected_val):
         dag = DAG(dag_id='simple_dag', params=val)
         BaseOperator(task_id='simple_task', dag=dag, start_date=datetime(2019, 8, 1))
 
-        serialized_dag = SerializedDAG.to_dict(dag)
+        serialized_dag_json = SerializedDAG.to_json(dag)
+
+        serialized_dag = json.loads(serialized_dag_json)
```
Comment (kaxil) [inline tests/serialization/test_dag_serialization.py:740]: Any reason we had to change this?

File: tests/serialization/test_dag_serialization.py [hunk 10]
```diff
@@ -1433,29 +1461,32 @@ def test_serialized_objects_are_sorted(self, object_to_serialized, expected_outp
         assert serialized_obj == expected_output
 
     def test_params_upgrade(self):
+        """when pre-2.2.0 param (i.e. primitive) is deserialized we convert to Param"""
         serialized = {
             "__version": 1,
             "dag": {
                 "_dag_id": "simple_dag",
-                "fileloc": __file__,
+                "fileloc": '__file__',
```
Comment (kaxil) [inline tests/serialization/test_dag_serialization.py]: Why this change?

File: airflow/models/param.py [hunk 11]
```diff
@@ -49,18 +50,37 @@ class Param:
...
+        """
+        try:
+            json.dumps(value)
+        except TypeError:
+            warnings.warn(
+                "The use of non-json-serializable params is deprecated and will be removed in a"
+                " future release",
+                DeprecationWarning,
+                stacklevel=2,
+            )
```
Comment (kaxil) [inline airflow/models/param.py]: Let's separate deprecation from this PR -- so we can think more about it.

File: tests/serialization/test_dag_serialization.py (hunk 10 above)
Comment (dstandish) [reply tests/serialization/test_dag_serialization.py]: this value actually does not have any impact on the test.  it could be `/hello/world.py`  
but when you  put `__file__` you  can't run the bit of code interactively.

```python
serialized = {
    "__version": 1,
    "dag": {
        "_dag_id": "simple_dag",
        "fileloc": __file__,
        "tasks": [],
        "timezone": "UTC",
        "params": {"none": None, "str": "str", "dict": {"a": "b"}},
    },
}
dag = SerializedDAG.from_dict(serialized)
Traceback (most recent call last):
  File "/Users/dstandish/.virtualenvs/local/lib/python3.8/site-packages/IPython/core/interactiveshell.py", line 3444, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-2-4ee74d1eaf6b>", line 5, in <module>
    "fileloc": __file__,
NameError: name '__file__' is not defined
```

File: tests/serialization/test_dag_serialization.py (hunk 9 above)
Comment (dstandish) [reply tests/serialization/test_dag_serialization.py:740]: by calling `to_json` we make this test stricter.  the full serialization process must convert not only to a dictionary (which is python, thus can handle arbitrary data types) but also ultimately must serialize to json.  

the test previously did not convert to json; only to dict and back.  so without making this change, it would not catch the `set` issue.

so here we first go all the way to json before deserilaizing.  

in other words, previously the round trip wasn't really a full round trip and that's what this change fixes.

File: airflow/models/param.py (hunk 11 above)
Comment (dstandish) [reply airflow/models/param.py]: ok sure.

Comment (dstandish) [reply airflow/models/param.py]: done

**Review by kaxil — APPROVED** [review]

**Review by jedcunningham — APPROVED** [review]
